### PR TITLE
[luci/pass] Extend FuseInstanceNormPass with keras support

### PR DESCRIPTION
**Pull request by mbencer**
This commit adds a new pattern of InstanceNorm fusing. It's based on the tf.keras.layers.GroupNormalization(groups=-1) layer from keras library.

ONE-DCO-1.0-Signed-off-by: Mateusz Bencer m.bencer@partner.samsung.com

Issue: https://github.com/Samsung/ONE/issues/16347

### CODE REVIEW

**Comment by mbencer** [comment]
image

File: compiler/luci/pass/src/FuseInstanceNormPass.cpp [hunk 1]
```diff
@@ -751,6 +869,81 @@ template <> bool InstanceNormPattern::match<InstanceNormPattern::PatternVersion:
   return true;
 }
 
+template <> bool InstanceNormPattern::match<InstanceNormPattern::PatternVersion::Version_7>()
+{
+  add_as_terminal = dynamic_cast<luci::CircleAdd *>(reshape_as_terminal->tensor());
```
Comment (mbencer) [inline compiler/luci/pass/src/FuseInstanceNormPass.cpp:873]: 14-15

File: compiler/luci/pass/src/FuseInstanceNormPass.cpp [hunk 2]
```diff
@@ -751,6 +869,81 @@ template <> bool InstanceNormPattern::match<InstanceNormPattern::PatternVersion:
...
+  CHECK_OR_FALSE(add_as_terminal);
+
+  CHECK_OR_FALSE(luci::fill(&mul_as_scaled_ifm, &add_neg_mul).with_commutative_args_of(add_as_terminal));
```
Comment (mbencer) [inline compiler/luci/pass/src/FuseInstanceNormPass.cpp]: (11,13)-14

File: compiler/luci/pass/src/FuseInstanceNormPass.cpp [hunk 3]
```diff
@@ -751,6 +869,81 @@ template <> bool InstanceNormPattern::match<InstanceNormPattern::PatternVersion:
...
+  CHECK_OR_FALSE(luci::fill(&reshape_of_ifm, &mul_gamma).with_commutative_args_of(mul_as_scaled_ifm));
```
Comment (mbencer) [inline compiler/luci/pass/src/FuseInstanceNormPass.cpp]: (9,2)-11

File: compiler/luci/pass/src/FuseInstanceNormPass.cpp [hunk 4]
```diff
@@ -751,6 +869,81 @@ template <> bool InstanceNormPattern::match<InstanceNormPattern::PatternVersion:
...
+
+  mul_as_scaled_mean = dynamic_cast<luci::CircleMul *>(add_neg_mul->x());
```
Comment (mbencer) [inline compiler/luci/pass/src/FuseInstanceNormPass.cpp:881]: 12-13

File: compiler/luci/pass/src/FuseInstanceNormPass.cpp [hunk 5]
```diff
@@ -751,6 +869,81 @@ template <> bool InstanceNormPattern::match<InstanceNormPattern::PatternVersion:
...
+  CHECK_OR_FALSE(mul_as_scaled_mean);
+
+  neg_mean = dynamic_cast<luci::CircleNeg *>(mul_as_scaled_mean->x());
```
Comment (mbencer) [inline compiler/luci/pass/src/FuseInstanceNormPass.cpp:884]: 10-12

File: compiler/luci/pass/src/FuseInstanceNormPass.cpp [hunk 6]
```diff
@@ -751,6 +869,81 @@ template <> bool InstanceNormPattern::match<InstanceNormPattern::PatternVersion:
...
+  CHECK_OR_FALSE(neg_mean);
+
+  luci::CircleMul *mul_gamma_should_be = nullptr;
+  luci::CircleNeg *neg_should_be = nullptr;
+
+  CHECK_OR_FALSE(luci::fill(&mul_gamma_should_be, &neg_should_be)
```
Comment (mbencer) [inline compiler/luci/pass/src/FuseInstanceNormPass.cpp]: (9,10)-12

File: compiler/luci/pass/src/FuseInstanceNormPass.cpp [hunk 7]
```diff
@@ -751,6 +869,81 @@ template <> bool InstanceNormPattern::match<InstanceNormPattern::PatternVersion:
...
+                   .with_commutative_args_of(mul_as_scaled_mean));
+
+
+  CHECK_OR_FALSE(mul_gamma == mul_gamma_should_be);
+  CHECK_OR_FALSE(neg_mean == neg_should_be);
+
+  mean_of_ifm = dynamic_cast<luci::CircleMean *>(neg_mean->x());
```
Comment (mbencer) [inline compiler/luci/pass/src/FuseInstanceNormPass.cpp:896]: 3-10

File: compiler/luci/pass/src/FuseInstanceNormPass.cpp [hunk 8]
```diff
@@ -751,6 +869,81 @@ template <> bool InstanceNormPattern::match<InstanceNormPattern::PatternVersion:
...
+  CHECK_OR_FALSE(mean_of_ifm);
+
+  luci::CircleReshape *reshape_of_ifm_should_be = nullptr;
+  reshape_of_ifm_should_be = dynamic_cast<luci::CircleReshape *>(mean_of_ifm->input());
+  CHECK_OR_FALSE(reshape_of_ifm_should_be == reshape_of_ifm);
```
Comment (mbencer) [inline compiler/luci/pass/src/FuseInstanceNormPass.cpp:901]: 2-3

File: compiler/luci/pass/src/FuseInstanceNormPass.cpp [hunk 9]
```diff
@@ -751,6 +869,81 @@ template <> bool InstanceNormPattern::match<InstanceNormPattern::PatternVersion:
...
+
+  ifm = reshape_of_ifm->tensor();
+  auto ifm_circle = loco::must_cast<luci::CircleNode *>(ifm);
+  CHECK_OR_FALSE(ifm_circle);
```
Comment (mbencer) [inline compiler/luci/pass/src/FuseInstanceNormPass.cpp:905]: 1-2

File: compiler/luci/pass/src/FuseInstanceNormPass.cpp [hunk 10]
```diff
@@ -751,6 +869,81 @@ template <> bool InstanceNormPattern::match<InstanceNormPattern::PatternVersion:
...
+
+  CHECK_OR_FALSE(ifm_circle->shape_status() == luci::ShapeStatus::VALID);
+  CHECK_OR_FALSE(ifm_circle->rank() == 4);
+  CHECK_OR_FALSE(ifm_circle->dim(3).known());
+  uint32_t ifm_channel_depth = ifm_circle->dim(3).value();
+
+  const_as_beta = dynamic_cast<luci::CircleConst *>(add_neg_mul->y());
+  CHECK_OR_FALSE(const_as_beta);
```
Comment (mbencer) [inline compiler/luci/pass/src/FuseInstanceNormPass.cpp:913]: const_as_beta-13

File: compiler/luci/pass/src/FuseInstanceNormPass.cpp [hunk 11]
```diff
@@ -751,6 +869,81 @@ template <> bool InstanceNormPattern::match<InstanceNormPattern::PatternVersion:
...
+  CHECK_OR_FALSE(is_1D_with_dummy_dim(const_as_beta, ifm_channel_depth));
+
+  CHECK_OR_FALSE(luci::fill(&rsqrt, &const_as_gamma).with_commutative_args_of(mul_gamma));
```
Comment (mbencer) [inline compiler/luci/pass/src/FuseInstanceNormPass.cpp:916]: 8-9

File: compiler/luci/pass/src/FuseInstanceNormPass.cpp [hunk 12]
```diff
@@ -751,6 +869,81 @@ template <> bool InstanceNormPattern::match<InstanceNormPattern::PatternVersion:
...
+  CHECK_OR_FALSE(is_1D_with_dummy_dim(const_as_gamma, ifm_channel_depth));
+
+  add_as_variance = dynamic_cast<luci::CircleAdd *>(rsqrt->x());
+  CHECK_OR_FALSE(add_as_variance);
```
Comment (mbencer) [inline compiler/luci/pass/src/FuseInstanceNormPass.cpp:920]: 7-8

File: compiler/luci/pass/src/FuseInstanceNormPass.cpp [hunk 13]
```diff
@@ -751,6 +869,81 @@ template <> bool InstanceNormPattern::match<InstanceNormPattern::PatternVersion:
...
+
+  CHECK_OR_FALSE(luci::fill(&mean_as_variance, &const_as_epsilon).with_commutative_args_of(add_as_variance));
```
Comment (mbencer) [inline compiler/luci/pass/src/FuseInstanceNormPass.cpp]: (6,const_as_epsilon)-7

File: compiler/luci/pass/src/FuseInstanceNormPass.cpp [hunk 14]
```diff
@@ -751,6 +869,81 @@ template <> bool InstanceNormPattern::match<InstanceNormPattern::PatternVersion:
...
+  CHECK_OR_FALSE(mean_as_variance);
+
+  CHECK_OR_FALSE(const_as_epsilon->dtype() == loco::DataType::FLOAT32);
+  // TODO Support regarding broadcast
+  CHECK_OR_FALSE(const_as_epsilon->size<loco::DataType::FLOAT32>() == 1);
+
+  square = dynamic_cast<luci::CircleSquare *>(mean_as_variance->input());
+  CHECK_OR_FALSE(square);
```
Comment (mbencer) [inline compiler/luci/pass/src/FuseInstanceNormPass.cpp:931]: 5-6

File: compiler/luci/pass/src/FuseInstanceNormPass.cpp [hunk 15]
```diff
@@ -751,6 +869,81 @@ template <> bool InstanceNormPattern::match<InstanceNormPattern::PatternVersion:
...
+
+  sub_2 = dynamic_cast<luci::CircleSub *>(square->x());
+  CHECK_OR_FALSE(sub_2);
```
Comment (mbencer) [inline compiler/luci/pass/src/FuseInstanceNormPass.cpp:934]: 4-5

File: compiler/luci/pass/src/FuseInstanceNormPass.cpp [hunk 16]
```diff
@@ -751,6 +869,81 @@ template <> bool InstanceNormPattern::match<InstanceNormPattern::PatternVersion:
...
+  
+  auto mean_of_ifm_should_be = dynamic_cast<luci::CircleMean *>(sub_2->y());
+  CHECK_OR_FALSE(mean_of_ifm == mean_of_ifm_should_be);
```
Comment (mbencer) [inline compiler/luci/pass/src/FuseInstanceNormPass.cpp:937]: 3-4

File: compiler/luci/pass/src/FuseInstanceNormPass.cpp [hunk 17]
```diff
@@ -751,6 +869,81 @@ template <> bool InstanceNormPattern::match<InstanceNormPattern::PatternVersion:
...
+
+  auto reshape_of_ifm_should_be_2 = dynamic_cast<luci::CircleReshape *>(sub_2->x());
+  CHECK_OR_FALSE(reshape_of_ifm_should_be_2 == reshape_of_ifm);
```
Comment (mbencer) [inline compiler/luci/pass/src/FuseInstanceNormPass.cpp:940]: 2-4

File: compiler/luci/pass/src/FuseInstanceNormPass.cpp [hunk 18]
```diff
@@ -25,20 +25,82 @@
...
 // Helper to check detail
 
-/// @return true  When node has shape of '1 x .. x 1 x depth'
+#define CHECK_OR_FALSE(condition) \
+  if (not(condition))             \
+    return false;
+
+/// @return true  When node has shape with one dim other than `1`  (like '1 x .. x 1 x depth' or '1
+/// x .. x depth' x 1)
 bool is_1D_with_dummy_dim(luci::CircleConst *node, uint32_t depth)
```
Comment (tomdol) [inline compiler/luci/pass/src/FuseInstanceNormPass.cpp]: The name of this helper is really confusing. Can we rename it to something like `is_unsqueezed_1D`? Given the current name I can't really figure out what a dummy dim means in the context of something that is supposed to be 1D. And this is how I read the name of this function.

File: compiler/luci/pass/src/FuseInstanceNormPass.cpp [hunk 19]
```diff
@@ -25,20 +25,82 @@
...
+
+/// @return true if the provided begin_reshape Reshape op adds `1` dimension
+///         and terminal_reshape Reshape op removes it (the result is neutral for further
+///         processing)
+bool is_unsqueeze_squeeze_pair(luci::CircleReshape *begin_reshape,
+                               luci::CircleReshape *terminal_reshape)
+{
+  auto const begin_ifm = dynamic_cast<luci::CircleNode *>(begin_reshape->tensor());
+  CHECK_OR_FALSE(begin_ifm);
+  auto const begin_ofm = loco::must_cast<luci::CircleNode *>(begin_reshape);
```
Comment (tomdol) [inline compiler/luci/pass/src/FuseInstanceNormPass.cpp]: Is this cast required? Since `CircleReshape` is a subclass of `CircleNode` I suppose it should be possible to access the base class' member functions used below. The same goes for the `terminal_ofm` node in this scope.

File: compiler/luci/pass/src/FuseInstanceNormPass.cpp [hunk 20]
```diff
@@ -25,20 +25,82 @@
...
+  CHECK_OR_FALSE(begin_ofm);
+
+  // check last axis
+  CHECK_OR_FALSE((begin_ifm->rank() + 1) == begin_ofm->rank());
```
Comment (tomdol) [inline compiler/luci/pass/src/FuseInstanceNormPass.cpp]: Shouldn't those ranks be equal? It seems that this check verifies that the input to the first reshape and the output of the last reshape match in terms of the dimensionality. If so, why do we add a 1 here?

File: compiler/luci/pass/src/FuseInstanceNormPass.cpp [hunk 21]
```diff
@@ -751,6 +872,83 @@ template <> bool InstanceNormPattern::match<InstanceNormPattern::PatternVersion:
   return true;
 }
 
+template <> bool InstanceNormPattern::match<InstanceNormPattern::PatternVersion::Version_7>()
+{
+  add_as_terminal = dynamic_cast<luci::CircleAdd *>(reshape_as_terminal->tensor());
+  CHECK_OR_FALSE(add_as_terminal);
+
+  CHECK_OR_FALSE(
```
Comment (mbencer) [inline compiler/luci/pass/src/FuseInstanceNormPass.cpp:876]: (11,13)-14

File: compiler/luci/pass/src/FuseInstanceNormPass.cpp [hunk 22]
```diff
@@ -751,6 +872,83 @@ template <> bool InstanceNormPattern::match<InstanceNormPattern::PatternVersion:
...
+    luci::fill(&mul_as_scaled_ifm, &add_neg_mul).with_commutative_args_of(add_as_terminal));
+  CHECK_OR_FALSE(
```
Comment (mbencer) [inline compiler/luci/pass/src/FuseInstanceNormPass.cpp:878]: (2,9)-11

File: compiler/luci/pass/src/FuseInstanceNormPass.cpp [hunk 23]
```diff
@@ -751,6 +872,83 @@ template <> bool InstanceNormPattern::match<InstanceNormPattern::PatternVersion:
...
+  mul_as_scaled_mean = dynamic_cast<luci::CircleMul *>(add_neg_mul->x());
+  CHECK_OR_FALSE(mul_as_scaled_mean);
+
+  neg_mean = dynamic_cast<luci::CircleNeg *>(mul_as_scaled_mean->x());
+  CHECK_OR_FALSE(neg_mean);
+
+  luci::CircleMul *mul_gamma_should_be = nullptr;
+  luci::CircleNeg *neg_should_be = nullptr;
+
+  CHECK_OR_FALSE(
```
Comment (mbencer) [inline compiler/luci/pass/src/FuseInstanceNormPass.cpp:890]: (9,10)-12

File: compiler/luci/pass/src/FuseInstanceNormPass.cpp [hunk 24]
```diff
@@ -751,6 +872,83 @@ template <> bool InstanceNormPattern::match<InstanceNormPattern::PatternVersion:
...
+  CHECK_OR_FALSE(is_1D_with_dummy_dim(const_as_beta, ifm_channel_depth));
+
+  CHECK_OR_FALSE(luci::fill(&rsqrt, &const_as_gamma).with_commutative_args_of(mul_gamma));
+  CHECK_OR_FALSE(is_1D_with_dummy_dim(const_as_gamma, ifm_channel_depth));
+
+  add_as_variance = dynamic_cast<luci::CircleAdd *>(rsqrt->x());
+  CHECK_OR_FALSE(add_as_variance);
+
+  CHECK_OR_FALSE(
+    luci::fill(&mean_as_variance, &const_as_epsilon).with_commutative_args_of(add_as_variance));
```
Comment (mbencer) [inline compiler/luci/pass/src/FuseInstanceNormPass.cpp:923]: (6, const_as_epsilon)-7

File: compiler/luci/pass/src/FuseInstanceNormPass.cpp [hunk 25]
```diff
@@ -751,6 +872,83 @@ template <> bool InstanceNormPattern::match<InstanceNormPattern::PatternVersion:
...
+  CHECK_OR_FALSE(
+    luci::fill(&reshape_of_ifm, &mul_gamma).with_commutative_args_of(mul_as_scaled_ifm));
+
+  mul_as_scaled_mean = dynamic_cast<luci::CircleMul *>(add_neg_mul->x());
+  CHECK_OR_FALSE(mul_as_scaled_mean);
+
+  neg_mean = dynamic_cast<luci::CircleNeg *>(mul_as_scaled_mean->x());
+  CHECK_OR_FALSE(neg_mean);
+
+  luci::CircleMul *mul_gamma_should_be = nullptr;
```
Comment (tomdol) [inline compiler/luci/pass/src/FuseInstanceNormPass.cpp:887]: Not sure about the naming of those variables. Are they supposed to denote that some particular nodes exist? If so I suggest something like `mul_gamma_exists`.

File: compiler/luci/pass/src/FuseInstanceNormPass.cpp (hunk 20 above)
Comment (mbencer) [reply compiler/luci/pass/src/FuseInstanceNormPass.cpp]: I'm comparing here rank of InstanceNorm ifm with output of the first Reshape (which plays Unsqueeze role here):
image

File: compiler/luci/pass/src/FuseInstanceNormPass.cpp (hunk 19 above)
Comment (mbencer) [reply compiler/luci/pass/src/FuseInstanceNormPass.cpp]: right, it was some leftovers from previous version, thank you. Fixed

File: compiler/luci/pass/src/FuseInstanceNormPass.cpp (hunk 18 above)
Comment (mbencer) [reply compiler/luci/pass/src/FuseInstanceNormPass.cpp]: I see, How about `is_flat_tensor`?

File: compiler/luci/pass/src/FuseInstanceNormPass.cpp (hunk 25 above)
Comment (mbencer) [reply compiler/luci/pass/src/FuseInstanceNormPass.cpp:887]: I'm checking here if `mul_gamma` `ofm` indicated by "green" edge (calculated earlier) is the same like "red" edge here (just by an address check).
image

I also have doubts about `*_should_be` name but I've tried to be consistent with other passes. IMO it should be refactored separately (at least for `FuseInstanceNorm.cpp`)

File: compiler/luci/pass/src/FuseInstanceNormPass.cpp (hunk 20 above)
Comment (tomdol) [reply compiler/luci/pass/src/FuseInstanceNormPass.cpp]: Right, my bad. Please ignore this question.

File: compiler/luci/pass/src/FuseInstanceNormPass.cpp (hunk 18 above)
Comment (mbencer) [reply compiler/luci/pass/src/FuseInstanceNormPass.cpp]: @tomdol  After rethinking `is_unsqueezed_1D` seems to be most readable name here
